Optimize Docker image size by removing unused dependencies and creating runtime-optimized container.

Jayssgss opened the pull request:
Reduced pyproject.toml from 95+ dependencies to 5 core runtime dependencies + 7 dev dependencies,  container size reduction to 259MB.

Implemented dependency groups for development/testing while maintaining full CI compatibility. 

Created optimized Dockerfile that installs only runtime dependencies and integrates environment credentials, resulting in a production-ready container with all 16 MCP tools functional.

Jayssgss committed:
Reducing the container size by using runtime dependencies only and removing unused dependencies from the pyproject.toml

Jayssgss committed:
Merge remote-tracking branch 'upstream/main' into INSTA-51198

Jayssgss committed:
Making ruff available as a dependency for ci builds

Jayssgss committed:
Resolving dependency issue for ruff in ci.yml and fixing test failures

arthurdm reviewed (changes requested):
Thanks for working on this @Jayssgss.  First pass observations: 

- API keys and tokens are hardcoded in the Dockerfile (INSTANA_API_TOKEN, WATSONX_API_KEY, etc.)
  -   These should be passed as environment variables at runtime or through a secure secret management system
- The `apt-get` install command in the runtime stage is malformed (has && but no package list)
- The builder stage installs gcc but doesn't clean it up, increasing the image size
- No non-root user is created (running as root is a security risk)
- No explicit `--no-install-recommends` in `apt-get`
- The health check assumes the application will be running on localhost, but in a container, this might not be reliable

Jayssgss committed:
optimize Docker image size and implement security best practices

- Reduce dependencies from 95+ to 20 essential runtime packages
- Implement two-file dependency strategy (dev vs production)
- Add multi-stage Docker build with security hardening
- Remove hardcoded secrets, use environment variables
- Create non-root user (mcpuser) for container security
- Add health checks and proper container networking
- Optimize apt-get with --no-install-recommends flag
- Reduce image size by ~70-80% (266MB vs 1-2GB+)
- Add comprehensive Docker deployment documentation
- Include Kubernetes and Docker Compose examples

Jayssgss commented:
Hey @arthurdm ,

**Security Issues Fixed:**

1. **Hardcoded secrets removed**: All API keys and tokens are now passed via environment variables at runtime
2. **Non-root user implemented**: Container now runs as `mcpuser` instead of root
3. **Multi-stage build optimized**: Build tools (gcc, g++) are only in builder stage, not copied to runtime
4. **Apt-get flags added**: Using `--no-install-recommends` to minimize package installation
5. **Health check fixed**: Now uses `127.0.0.1` instead of `localhost` for container networking

## **Additional Improvements:**

- **Dependency optimization**: Reduced from 95+ dependencies to 20 essential runtime dependencies
- **Image size reduction**: ~70-80% smaller images (266MB vs 1-2GB+)
- **Two-file strategy**: `pyproject.toml` for development, `pyproject-runtime.toml` for production
- **Comprehensive documentation**: Added Docker deployment section to README.md

##  **Technical Details:**

- **Builder stage**: Installs only runtime dependencies from `pyproject-runtime.toml`
- **Runtime stage**: Minimal production image with non-root user and health checks
- **Security**: No secrets in image, proper user permissions, optimized base image
- **Production ready**: Includes Docker Compose and Kubernetes deployment examples

gyliu513 commented on README.md:
````diff
@@ -920,8 +920,226 @@ The azureManagedHSM plugin provides three metrics that can help monitor service
 3. Overall Service Availability: This metric measures the availability of the service.
 ```
 
+## Docker Deployment
````
Can you run `doctoc` to update toc?

Jayssgss replied:
I have added the indexing for the docker deployment documentation using doctoc now.
Thanks

gyliu513 commented on src/application/application_settings.py:
```diff
@@ -19,25 +19,19 @@
 
 # Import the necessary classes from the SDK
 try:
-    from instana_client.api.application_settings_api import (
```
Can you do this in another PR if this is not related to the dockerfile?

Jayssgss replied:
Actually this was to resolve some ci build error, the package was somehow not being imported from instana client sdk, so had to follow this way of importing it

Jayssgss committed:
Adding Doctoc Indexing for Docker Documentation in README
